spec/lib/reports/sales_tax_totals_by_order_spec.rb

```diff
@@ -0,0 +1,258 @@
+    variant.update!(tax_category_id: tax_category.id)
+
+    order.update!(
+      number: 'ORDER_NUMBER_1',
+      order_cycle_id: order_cycle.id,
+      ship_address_id: ship_address.id,
+      customer_id: customer1.id,
+      email: 'order1@example.com'
+    )
+    order.line_items.create(variant:, quantity: 1, price: 100)
+
+    # the enterprise fees can be known only when the user selects the variants
+    # we'll need to create them by calling recreate_all_fees!
+    order.recreate_all_fees!
+  end
+
+  describe "#filtered_tax_rate_total" do
+    let(:query_row) do
+      [
+        [country_tax_rate.id, order.id],
+        order
+      ]
+    end
+
+    # TODO see if we can simplify setup
+    # sould be able to only create the adjustment we need, or maybe mock them ?
+    it "returns tax amount filtered by tax rate in query_row" do
+      OrderWorkflow.new(order).complete!
+
+      filtered_tax_total = report.filtered_tax_rate_total(query_row)
+
+      expect(filtered_tax_total).not_to eq(order.total_tax)
+
+      # 10 % of 10.00 shipment cost + 10 % of 100.00 line item price
+      expect(filtered_tax_total).to eq(0.1 + 1)
+    end
+  end
+
+  describe "#tax_rate_total" do
+    let(:query_row) do
+      [
+        [state_tax_rate.id, order.id],
+        order
+      ]
+    end
+
+    it "returns the tax amount filtered by tax rate in the query_row" do
+      OrderWorkflow.new(order).complete!
+
+      tax_total = report.tax_rate_total(query_row)
+
+      expect(tax_total).not_to eq(order.total_tax)
+
+      # 20 % of 10.00 shipment cost + 20 % of 100.00 line item price
+      expect(tax_total).to eq(0.2 + 2)
+    end
+
+    context "with a voucher" do
+      let(:voucher) do
+        create(:voucher_flat_rate, code: 'some_code', enterprise: order.distributor, amount: 10)
+      end
+
+      it "returns the tax amount adjusted with voucher tax discount" do
+        add_voucher(order, voucher)
+
+        tax_total = report.tax_rate_total(query_row)
+
+        # 20 % of 10.00 shipment cost + 20 % of 100.00 line item price - voucher tax
+        expect(tax_total).to eq(0.2 + 2 - 0.29)
+      end
+    end
+  end
+
+  describe "#total_excl_tax" do
+    let(:query_row) do
+      [
+        [state_tax_rate.id, order.id],
+        order
+      ]
+    end
+
+    it "returns the total excluding tax specified in query_row" do
+      OrderWorkflow.new(order).complete!
+
+      total = report.total_excl_tax(query_row)
+
+      # order.total - order.total_tax
+      expect(total).to eq(113.3 - 3.3)
+    end
+
+    context "with a voucher" do
+      let(:voucher) do
+        create(:voucher_flat_rate, code: 'some_code', enterprise: order.distributor, amount: 10)
+      end
+
+      it "returns the total exluding tax and indcluding voucher tax discount" do
+        add_voucher(order, voucher)
+
+        total = report.total_excl_tax(query_row)
+
+        # order.total - order.total_tax + voucher tax
+        expect(total).to eq(103.3 - 3.3 + 0.29)
```

Comment:
At first I think I understand. The order is 113.3 including tax. But the voucher discounts the order by 10, meaning the customer only pays 103.3. Tax is only applicable to what was paid, so only now there is only 3.01 tax (3.3-0.29). 

Hmm I could be wrong, but I think this comment is misleading. I think the model method `order.total_tax` would return the discounted `3.01` (just like `order.total` returns the discounted `103.3`). 
Is this more correct? (I'm not sure, the numbers still seem wrong..) 

```suggestion
        # discounted order total - discounted order tax
        expect(total).to eq((113.3 - 10) - (3.3 - 0.29))
```

Reply:
You are right it's a bit misleading, it seems to make sense when I wrote it but not so much now.
FYI `order.total_tax` doesn't returned the discounted tax because :
1. voucher tax part isn't always modelled as an adjustment
2. when modelled as an adjustment it's not considered a "Tax adjustment"

I have been a bit reluctant to fix it, because I am not sure what the implication would be in other part of OFN.